### PR TITLE
Pool Manager: limit, pause and cancel requests

### PR DESCRIPTION
**Pull request by vixez**
This is a PR related to https://github.com/CodingAleCR/http_interceptor/issues/92

The goal is to limit the amount of simultaneous requests, and pause requests when a token is being updated.
It is also possible to cancel active requests.

`PoolManager` can handle both of these.
I have added an explanation to `README.md` and updated `CHANGELOG.md` as well.

### Pool Manager  
  
Using a `PoolManager` allows you to set the maximum amount of simultaneous requests, and pause  
requests during token updates.  
  
Create a `PoolManager` and assign it to an `InterceptedClient`:  
  
```dart 
final PoolManager poolManager = PoolManager(    
  maxActiveConnections: 8, 
);  
  
InterceptedClient.build(    
  poolManager: poolManager, 
);  
```   
In the above example only 8 requests can be active at once. By default this is 32.  

#### Skip the pool

It is possible to skip a pool that is active, on hold, or blocked by a token update.
This can be useful to, for example, update a token in a `RetryPolicy` while the pool is on hold.
If the `InterceptedClient.kSkipPoolHeader` header is found in the request the pool will be skipped, so the request is executed immediately.

To your headers, add for example:
```dart
headers[InterceptedClient.kSkipPoolHeader] = 'true';
```

The value of the header does not matter and the `InterceptedClient.kSkipPoolHeader` is filtered out of the request before being sent to the server.
  
#### Refreshing a token  
  
If you use a `RetryPolicy` you can use the `PoolManager` to put new requests on hold until  
the `RetryPolicy` has completed. Use `requestUpdateToken` to let the `PoolManager` know you want to  
update the token and put new requests on hold. Once you have updated the token,  
use `releaseUpdateToken` to resume requests.  
  
Be sure to add logic (through an interceptor) to update queued requests with the new token you got  
in the `RetryPolicy`.  
  
`poolManager` in the example should be a reference to where you stored the `PoolManager` you  
assigned to `InterceptedClient`.  
  
```dart  
class TokenRetryPolicy extends RetryPolicy {  
  @override Future<bool> shouldAttemptRetryOnResponse(ResponseData response) async { 
    if (response.statusCode == 401) { 
     // Need to get a new token 
     await poolManager.requestUpdateToken();  
     
     /*  --- YOUR TOKEN REFRESH LOGIC --- */  
     
     poolManager.releaseUpdateToken(); return true; } 
     return false; 
   }
 }  
```

#### Cancel active requests
You can cancel all requests that have not yet returned a response. They will return with `RequestCancelledException`.
This is done through the `PoolManager` with `cancelRequests` (optionally releasing token updates with the `forceReleaseUpdateToken`) parameter.

### CODE REVIEW

**Comment by CodingAleCR** [comment]
Hey, this looks amazing, I will continue testing it out with the new 2.0.0 APIs, thank you for all the hard work you have put into this!

**Comment by vixez** [comment]
You're welcome. Let me know if there is anything I need to change/can do.

**Comment by vixez** [comment]
I have updated this to work with the 2.0.0 beta.

**Comment by fnicastri** [comment]
Any update on this @CodingAleCR ? Looking forward to use this on my app(almost ready to be released) app!

Thanks for this awesome lib!

**Comment by CodingAleCR** [comment]
Not just yet, I have been trying to avoid adding significant new features while beta testing the new API that adds support for different types of requests and responses. 

However, I do appreciate the patience and the hard work been put in by the different contributors and interested supporters 🎉

**Comment by fnicastri** [comment]
@CodingAleCR @vixez  Thanks!

**Comment by vixez** [comment]
Added the ability to cancel all requests that have not yet returned a response

**Comment by vixez** [comment]
That could be an option as well, yes 🤔

**Comment by LucPham** [comment]
waiting!!

**Comment by fnicastri** [comment]
🤷‍♂️

File: lib/http_interceptor.dart [hunk 1]
```diff
@@ -1,6 +1,7 @@
 library http_interceptor;
 
-export 'package:http/http.dart';
```
Comment (CodingAleCR) [inline lib/http_interceptor.dart:3]: I think it's best if we expose `http` package along with this package so we can avoid having to install manually `http` when using `http_interceptor`.

**Review by CodingAleCR — CHANGES_REQUESTED** [review]
I've explored other options but none that I like as alternatives, this has the most potential and stability. Would you be willing to update the PR with the latest changes?

**Comment by shinayser** [comment]
Is this feature still be implemented or its abbandoned?

**Comment by CodingAleCR** [comment]
I asked for a couple of updates. When done it should be good to go

**Comment by vixez** [comment]
@shinayser I have rewritten major parts of our API code for the app, making this no longer needed.

**Pull request closed without merging**